### PR TITLE
Fix HASKEY BigInteger index overflow

### PR DESCRIPTION
**Pull request by Jim8y**
This pull request enhances the safety and correctness of array, buffer, and byte string index handling in the Neo VM by improving overflow and underflow checks and ensuring consistent type usage for indices. It also adds new test cases to verify these edge conditions.

Improvements to index validation and safety:

* Updated index validation logic in `HasKey` for `VMArray`, `Buffer`, and `ByteString` to check both lower and upper bounds using appropriate engine limits, preventing overflow and underflow errors. [[1]](diffhunk://#diff-753e88c77d252d79a5b1ca10bfe873789f6048a21fdfe5d344de0bed6ee365cbL283-R285) [[2]](diffhunk://#diff-753e88c77d252d79a5b1ca10bfe873789f6048a21fdfe5d344de0bed6ee365cbL299-R309)
* Standardized index handling in `PickItem`, `SetItem`, and `Remove` methods to use `BigInteger` for indices and explicitly cast to `int` only when accessing collections, improving type safety and clarity. [[1]](diffhunk://#diff-753e88c77d252d79a5b1ca10bfe873789f6048a21fdfe5d344de0bed6ee365cbL378-R378) [[2]](diffhunk://#diff-753e88c77d252d79a5b1ca10bfe873789f6048a21fdfe5d344de0bed6ee365cbL394-R402) [[3]](diffhunk://#diff-753e88c77d252d79a5b1ca10bfe873789f6048a21fdfe5d344de0bed6ee365cbL447-R447) [[4]](diffhunk://#diff-753e88c77d252d79a5b1ca10bfe873789f6048a21fdfe5d344de0bed6ee365cbL460-R465) [[5]](diffhunk://#diff-753e88c77d252d79a5b1ca10bfe873789f6048a21fdfe5d344de0bed6ee365cbL517-R517)

Testing improvements:

* Added new test cases in `HASKEY.json` to verify correct FAULT behavior when array indices overflow or underflow, ensuring these edge cases are properly handled.

### CODE REVIEW

File: src/Neo.VM/JumpTable/JumpTable.Compound.cs [hunk 1]
```diff
@@ -280,10 +280,15 @@ public virtual void HasKey(ExecutionEngine engine, Instruction instruction)
...
+                    var indexBI = key.GetInteger();
+                    if (indexBI.Sign < 0)
+                        throw new InvalidOperationException($"The negative index {indexBI} is invalid for OpCode.{instruction.OpCode}.");
+                    if (indexBI > int.MaxValue)
+                    {
+                        engine.Push(false);
+                        break;
+                    }
+                    var index = (int)indexBI;
                     engine.Push(index < array.Count);
```
Comment (ajara87) [inline src/Neo.VM/JumpTable/JumpTable.Compound.cs]:
```suggestion
                    engine.Push(indexBI < array.Count);
```
It's possible to compare directly without previous cast

File: src/Neo.VM/JumpTable/JumpTable.Compound.cs [hunk 2]
```diff
@@ -296,20 +301,30 @@ public virtual void HasKey(ExecutionEngine engine, Instruction instruction)
...
+                    var indexBI = key.GetInteger();
+                    if (indexBI.Sign < 0)
+                        throw new InvalidOperationException($"The negative index {indexBI} is invalid for OpCode.{instruction.OpCode}.");
+                    if (indexBI > int.MaxValue)
+                    {
+                        engine.Push(false);
+                        break;
+                    }
+                    var index = (int)indexBI;
                     engine.Push(index < buffer.Size);
```
Comment (ajara87) [inline src/Neo.VM/JumpTable/JumpTable.Compound.cs]:
```suggestion
                    engine.Push(indexBI < buffer.Size);
```

File: src/Neo.VM/JumpTable/JumpTable.Compound.cs [hunk 3]
```diff
@@ -296,20 +301,30 @@ public virtual void HasKey(ExecutionEngine engine, Instruction instruction)
...
+                    var indexBI = key.GetInteger();
+                    if (indexBI.Sign < 0)
+                        throw new InvalidOperationException($"The negative index {indexBI} is invalid for OpCode.{instruction.OpCode}.");
+                    if (indexBI > int.MaxValue)
+                    {
+                        engine.Push(false);
+                        break;
+                    }
+                    var index = (int)indexBI;
                     engine.Push(index < array.Size);
```
Comment (ajara87) [inline src/Neo.VM/JumpTable/JumpTable.Compound.cs]:
```suggestion
                    engine.Push(indexBI < array.Size);
```

**Review by ajara87 — APPROVED** [review]

**Review by shargon — COMMENTED** [review]
This require a Hardfork, and new Jump table

**Review by shargon — APPROVED** [review]
@Jim8y could you create the jump table in neo repository?

File: src/Neo.VM/JumpTable/JumpTable.Compound.cs (hunk 1 above)
Comment (erikzhang) [reply src/Neo.VM/JumpTable/JumpTable.Compound.cs]: Agree. And the condition `indexBI > int.MaxValue` can be removed.

**Review by shargon — APPROVED** [review]

**Comment by roman-khimov** [comment]
Why is it better than `OverflowException`? Realistically proper indexes are from 0 to 2047, so any code passing `MaxInt32+N` is broken in about the same way code passing -1 is.

**Review by erikzhang — DISMISSED** [review]

**Review by ajara87 — DISMISSED** [review]

**Review by Wi1l-B0t — DISMISSED** [review]

**Review by roman-khimov — DISMISSED** [review]
Wish there are more tests for it (that would spot the difference in behavior), but ok logically.

**Comment by roman-khimov** [comment]
Practically current JSON-based tests are very nice to me, at least they're easy to reuse for different VM implementations. Formal spec or model is a different story.

**Review by vncoelho — DISMISSED** [review]

**Comment by shargon** [comment]
Waiting for https://github.com/neo-project/neo/pull/4381

File: src/Neo.VM/JumpTable/JumpTable.Compound.cs [hunk 4]
```diff
@@ -280,10 +280,9 @@ public virtual void HasKey(ExecutionEngine engine, Instruction instruction)
             // For arrays, check if the index is within bounds and push the result onto the stack.
             case VMArray array:
                 {
-                    // TODO: Overflow and underflow checking needs to be done.
-                    var index = (int)key.GetInteger();
-                    if (index < 0)
-                        throw new InvalidOperationException($"The negative index {index} is invalid for OpCode.{instruction.OpCode}.");
+                    var index = key.GetInteger();
+                    if (index < 0 || index >= engine.Limits.MaxStackSize)
+                        throw new InvalidOperationException($"The index {index} is invalid for OpCode.{instruction.OpCode}.");
```
Comment (ajara87) [inline src/Neo.VM/JumpTable/JumpTable.Compound.cs]:
```suggestion
                        throw new InvalidOperationException($"The index {index} is invalid for OpCode {instruction.OpCode}.");
```

File: src/Neo.VM/JumpTable/JumpTable.Compound.cs [hunk 5]
```diff
@@ -296,20 +295,18 @@ public virtual void HasKey(ExecutionEngine engine, Instruction instruction)
             // For buffers, check if the index is within bounds and push the result onto the stack.
             case Buffer buffer:
                 {
-                    // TODO: Overflow and underflow checking needs to be done.
-                    var index = (int)key.GetInteger();
-                    if (index < 0)
-                        throw new InvalidOperationException($"The negative index {index} is invalid for OpCode.{instruction.OpCode}.");
+                    var index = key.GetInteger();
+                    if (index < 0 || index >= engine.Limits.MaxItemSize)
+                        throw new InvalidOperationException($"The index {index} is invalid for OpCode.{instruction.OpCode}.");
```
Comment (ajara87) [inline src/Neo.VM/JumpTable/JumpTable.Compound.cs]:
```suggestion
                        throw new InvalidOperationException($"The index {index} is invalid for OpCode {instruction.OpCode}.");
```

File: src/Neo.VM/JumpTable/JumpTable.Compound.cs [hunk 6]
```diff
@@ -296,20 +295,18 @@ public virtual void HasKey(ExecutionEngine engine, Instruction instruction)
...
             // For byte strings, check if the index is within bounds and push the result onto the stack.
             case ByteString array:
                 {
-                    // TODO: Overflow and underflow checking needs to be done.
-                    var index = (int)key.GetInteger();
-                    if (index < 0)
-                        throw new InvalidOperationException($"The negative index {index} is invalid for OpCode.{instruction.OpCode}.");
+                    var index = key.GetInteger();
+                    if (index < 0 || index >= engine.Limits.MaxItemSize)
+                        throw new InvalidOperationException($"The index {index} is invalid for OpCode.{instruction.OpCode}.");
```
Comment (ajara87) [inline src/Neo.VM/JumpTable/JumpTable.Compound.cs]:
```suggestion
                        throw new InvalidOperationException($"The index {index} is invalid for OpCode {instruction.OpCode}.");
```

**Review by Wi1l-B0t — APPROVED** [review]

**Review by cschuchardt88 — APPROVED** [review]

File: src/Neo.VM/JumpTable/JumpTable.Compound.cs [hunk 7]
```diff
@@ -280,10 +280,9 @@ public virtual void HasKey(ExecutionEngine engine, Instruction instruction)
             // For arrays, check if the index is within bounds and push the result onto the stack.
             case VMArray array:
                 {
-                    // TODO: Overflow and underflow checking needs to be done.
-                    var index = (int)key.GetInteger();
-                    if (index < 0)
-                        throw new InvalidOperationException($"The negative index {index} is invalid for OpCode.{instruction.OpCode}.");
+                    var index = key.GetInteger();
+                    if (index < 0 || index >= engine.Limits.MaxStackSize)
```
Comment (AnnaShaleva) [inline src/Neo.VM/JumpTable/JumpTable.Compound.cs]: `PICKITEM` handler validates `index` against array length instead of MaxStackSize. Let's unify this behaviour:
```suggestion
                    if (index < 0 || index >= array.Count)
```

Comment (AnnaShaleva) [reply src/Neo.VM/JumpTable/JumpTable.Compound.cs]: Ref. https://github.com/neo-project/neo-vm/blob/233a882a910779baa6b274459945cd859914958a/src/Neo.VM/JumpTable/JumpTable.Compound.cs#L376

File: src/Neo.VM/JumpTable/JumpTable.Compound.cs [hunk 8]
```diff
@@ -296,20 +295,18 @@ public virtual void HasKey(ExecutionEngine engine, Instruction instruction)
...
                 }
             // For byte strings, check if the index is within bounds and push the result onto the stack.
             case ByteString array:
                 {
-                    // TODO: Overflow and underflow checking needs to be done.
-                    var index = (int)key.GetInteger();
-                    if (index < 0)
-                        throw new InvalidOperationException($"The negative index {index} is invalid for OpCode.{instruction.OpCode}.");
+                    var index = key.GetInteger();
+                    if (index < 0 || index >= engine.Limits.MaxItemSize)
```
Comment (AnnaShaleva) [inline src/Neo.VM/JumpTable/JumpTable.Compound.cs:308]: As far as this one.

File: src/Neo.VM/JumpTable/JumpTable.Compound.cs (hunk 7 above)
Comment (shargon) [reply src/Neo.VM/JumpTable/JumpTable.Compound.cs]: Anna after merge your improvement, not it will throw an excwption instead of push false, not sure if it's the expected result

File: src/Neo.VM/JumpTable/JumpTable.Compound.cs [hunk 9]
```diff
@@ -281,7 +281,7 @@ public virtual void HasKey(ExecutionEngine engine, Instruction instruction)
             case VMArray array:
                 {
                     var index = key.GetInteger();
-                    if (index < 0 || index >= engine.Limits.MaxStackSize)
+                    if (index < 0 || index >= array.Count)
```
Comment (Wi1l-B0t) [inline src/Neo.VM/JumpTable/JumpTable.Compound.cs]: It's `HasKey`.

File: src/Neo.VM/JumpTable/JumpTable.Compound.cs (hunk 7 above)
Comment (AnnaShaleva) [reply src/Neo.VM/JumpTable/JumpTable.Compound.cs]: My bad, you're right, let's revert this change.

File: src/Neo.VM/JumpTable/JumpTable.Compound.cs [hunk 10]
```diff
@@ -280,10 +280,9 @@ public virtual void HasKey(ExecutionEngine engine, Instruction instruction)
             // For arrays, check if the index is within bounds and push the result onto the stack.
             case VMArray array:
                 {
-                    // TODO: Overflow and underflow checking needs to be done.
-                    var index = (int)key.GetInteger();
-                    if (index < 0)
-                        throw new InvalidOperationException($"The negative index {index} is invalid for OpCode.{instruction.OpCode}.");
+                    var index = key.GetInteger();
+                    if (index < 0 || index >= array.Count)
```
Comment (AnnaShaleva) [inline src/Neo.VM/JumpTable/JumpTable.Compound.cs]:
```suggestion
                    if (index < 0 || index >= engine.Limits.MaxItemSize)
```

File: src/Neo.VM/JumpTable/JumpTable.Compound.cs (hunk 7 above)
Comment (AnnaShaleva) [reply src/Neo.VM/JumpTable/JumpTable.Compound.cs]: Reverted.

File: src/Neo.VM/JumpTable/JumpTable.Compound.cs (hunk 9 above)
Comment (AnnaShaleva) [reply src/Neo.VM/JumpTable/JumpTable.Compound.cs]: Reverted.
